File: 2130-maximum-twin-sum-of-a-linked-list/2130-maximum-twin-sum-of-a-linked-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func pairSum(head *ListNode) int {
	arr := []int{}
	curr := head
	for curr != nil {
		arr = append(arr, curr.Val)
		curr = curr.Next
	}

	maxSum := 0
	for i := 0; i < len(arr)/2; i++ {
		maxSum = max(maxSum, arr[i]+arr[len(arr)-1-i])
	}

    return maxSum
}

// array, linked list
// time: O(n)
// space: O(n)

File: 2130-maximum-twin-sum-of-a-linked-list/2130-maximum-twin-sum-of-a-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        arr = []
        curr = head
        while curr:
            arr.append(curr.val)
            curr = curr.next

        max_sum = 0
        for i in range(len(arr) // 2):
            max_sum = max(max_sum, arr[i] + arr[len(arr) - 1 - i])
        
        return max_sum

# array, linked list
# time: O(n)
# space: O(n)